Add state flag file in K8s upgrades (#2702)

cicharka opened the pull request:
* it is based on solution under upgrade/tasks/kafka.yml
* added usage of update flag for masters and nodes - having single
  flag file for both was introducing malfunction (incomplete master
  upgrade flag was removed by tasks in nodes block)
* 'Switch apiserver address to local' task is now executed before the
  upgrade procedure in order to correctly fetch versions of particular
  masters - without it procedure for some masters could be skipped,
  because fetched version didn't match host it tried to upgrade
* separate _get_ and _set_ cluster version tasks
* introduce _initial_cluster_version_ used to evaluate need of
  an upgrade

Tested on cluster (Ubuntu machines) containing:
1. three k8s masters and three k8s nodes (HA) (upgrade from v1.0.1 to v1.3.0)

cicharka committed:
ansible: add state flag file in k8s upgrades

cicharka committed:
Changelog updated

atsikham commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -12,13 +12,22 @@
+        state: touch
+
     - name: Upgrade masters
-      when: cluster_version is version('v' + version, '<=')
```
Here and below we have single conditions, while common one for the whole block may be passed in different cases. Correct me if this is a wrong example/case: upgrade process at this point because of `kubelet_version`, but we don't need to execute this task.

cicharka replied:
> Have you checked creation of files for each incremental upgrade?

Yes, I did, works correctly.

> Here and below we have single conditions, while common one for the whole block may be passed in different cases. Correct me if this is a wrong example/case: upgrade process at this point because of kubelet_version, but we don't need to execute this task

Do you mean that we should separate **Upgrde masters** and **Upgrade nodes** tasks, so if only _kubelet_version_ is lower (_cluster_version_ is already desired) we will execute only **Upgrade nodes** task, and we will not execute **Upgrade masters**?

I was thinking about it and:
- based on fact that kubeadm clusters upgrade has to follow the workflow (primary cp node, additional cp nodes, worker nodes) and
- idempotency character of `kubeadm upgrade` command

I assumed that they can be coupled together. But I can easily change it so we will have previous conditions (I will revert lines 21 and 34 and I will just add flag condition to them)

atsikham replied:
@to-bar what is your opinion here? May be we could approve these changes with additional comments?

to-bar replied:
We run upgrade in serial mode (host by host). Let's consider a case when the first master had been upgraded successfully (the state flag file was removed) but upgrade failed on next master or on any node. The question is whether we may need to run upgrade tasks again on the first master (or 2nd master when failure was on 3rd one)? If we do not need to rerun upgrade tasks on any successfully upgraded node, I would vote for saving time and having separate conditions 😉

cicharka replied:
Cool, I will change it then :)

cicharka replied:
I moved PR to draft since I am not happy yet with solution provided for this part.

cicharka replied:
So finally I just moved state file check for masters and nodes. Inside upgrade of masters, the upgrade of first master and next master will be based on inventory check.

atsikham reviewed (changes requested):
Have you checked creation of files for each incremental upgrade?

to-bar commented on ansible/playbooks/roles/upgrade/defaults/main.yml:
```diff
@@ -19,3 +19,6 @@ opendistro_for_elasticsearch:
   upgrade_state_file_path: /etc/elasticsearch/epicli-upgrade-started.state
+
+kubernetes:
+  upgrade_flag_path: /tmp/kubernetes-upgrade-in-progress{{ ver }}.flag
```
```suggestion
  upgrade_state_file_path: /var/lib/epiphany/upgrade/state/kubernetes-{{ ver }}.uncompleted
```
Since `/tmp` is cleaned up after reboot I would use `/var/lib/epiphany/upgrade/state` instead. This location is ready to use since it's created by a task from `common` role which is run before `upgrade` playbook.

Changing name to `state_file` is for naming consistency, previously we used `lock_file` or `flag_file` but IMO `state_file` is more informative/intuitive 😉

cicharka replied:
Done!

cicharka committed:
Separate conditions for upgrade of masters and nodes

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

atsikham commented:
Just FYI, this PR should fix such issue that I met recently.

```
TASK [upgrade : k8s/install | Install kubeadm 1.20.12 package for Debian family] *************************************************************************
fatal: [ec2-18-212-2-237.compute-1.amazonaws.com]: FAILED! => {"cache_update_time": 1639014589, "cache_updated": true, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'kubeadm=1.20.12-00' 'kubelet=1.20.12-00'' failed: E: Packages were downgraded and -y was used without --allow-downgrades.\n", "rc": 100, "stderr": "E: Packages were downgraded and -y was used without --allow-downgrades.\n", "stderr_lines": ["E: Packages were downgraded and -y was used without --allow-downgrades."], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages will be DOWNGRADED:\n  kubeadm kubelet\n0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded.\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following packages will be DOWNGRADED:", "  kubeadm kubelet", "0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded."]}
```

cicharka commented:
> Just FYI, this PR should fix such issue that I met recently.
> 
> ```
> TASK [upgrade : k8s/install | Install kubeadm 1.20.12 package for Debian family] *************************************************************************
> fatal: [ec2-18-212-2-237.compute-1.amazonaws.com]: FAILED! => {"cache_update_time": 1639014589, "cache_updated": true, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'kubeadm=1.20.12-00' 'kubelet=1.20.12-00'' failed: E: Packages were downgraded and -y was used without --allow-downgrades.\n", "rc": 100, "stderr": "E: Packages were downgraded and -y was used without --allow-downgrades.\n", "stderr_lines": ["E: Packages were downgraded and -y was used without --allow-downgrades."], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages will be DOWNGRADED:\n  kubeadm kubelet\n0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded.\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following packages will be DOWNGRADED:", "  kubeadm kubelet", "0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded."]}
> ```

@atsikham I have not faced issue like that before, do you remember the way to reproduce it so I will be able to verify?

atsikham commented:
> > Just FYI, this PR should fix such issue that I met recently.
> > ```
> > TASK [upgrade : k8s/install | Install kubeadm 1.20.12 package for Debian family] *************************************************************************
> > fatal: [ec2-18-212-2-237.compute-1.amazonaws.com]: FAILED! => {"cache_update_time": 1639014589, "cache_updated": true, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'kubeadm=1.20.12-00' 'kubelet=1.20.12-00'' failed: E: Packages were downgraded and -y was used without --allow-downgrades.\n", "rc": 100, "stderr": "E: Packages were downgraded and -y was used without --allow-downgrades.\n", "stderr_lines": ["E: Packages were downgraded and -y was used without --allow-downgrades."], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages will be DOWNGRADED:\n  kubeadm kubelet\n0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded.\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following packages will be DOWNGRADED:", "  kubeadm kubelet", "0 upgraded, 0 newly installed, 2 downgraded, 0 to remove and 56 not upgraded."]}
> > ```
> 
> @atsikham I have not faced issue like that before, do you remember the way to reproduce it so I will be able to verify?

To reproduce that you need to run K8s upgrade on Ubuntu, interrupt that after the new kubeadm package is installed, and then run again.

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -12,14 +12,26 @@
+- name: Check for upgrade flag file
+  stat:
+    path: "{{ kubernetes.upgrade_state_file_path }}"
+  register: lock_file_status
```
```suggestion
- name: Check for upgrade state flag file
  stat:
    path: "{{ kubernetes.upgrade_state_file_path }}"
  register: upgrade_state_file_status
```
I would use upgrade state file for consistency and avoid using `lock` term since it is often used for preventing running multiple instances at the same time (e.g. by yum or apt).

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -12,14 +12,26 @@
     - name: Upgrade masters
-      when: cluster_version is version('v' + version, '<=')
+      when: >
+        lock_file_status.stat.exists or
```
```suggestion
        upgrade_state_file_status.stat.exists or
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -12,14 +12,26 @@
+        lock_file_status.stat.exists or
+        cluster_version is version('v' + version, '<')
       block:
+        - name: Create upgrade flag file
+          file:
+            path: "{{ kubernetes.upgrade_state_file_path }}"
+            state: touch
```
```suggestion
        - name: Create upgrade state file
          copy:
            dest: "{{ kubernetes.upgrade_state_file_path }}"
            content: Upgrade started
            mode: u=rw,g=r,o=
```
`touch` will update existing file each run so Ansible will report state as `changed`. `copy` will report `ok` when file already exists.

cicharka replied:
nice one, I will remember that!

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +42,31 @@
     - name: Upgrade nodes
-      when: kubelet_version is version('v' + version, '<=')
+      when: >
+        lock_file_status.stat.exists or
```
```suggestion
        upgrade_state_file_status.stat.exists or
```

cicharka committed:
* Improve naming
* Replace touch module with copy

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

cicharka committed:
* changed target hosts used for getting cluster version

With new approach we are going to check cluster version on
each master, so it will not skip upgrade masters block if first
master was already upgraded.

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

cicharka committed:
k8s upgrade improvement:

* added usage of update flag for masters and nodes - having single
  flag file for both was introducing malfunction (incomplete master
  upgrade flag was removed by tasks in nodes block)

* 'Switch apiserver address to local' task is now executed before the
  upgrade procedure in order to correctly fetch versions of particular
  masters - without it procedure for some masters could be skipped,
  because fetched version didn't match host it tried to upgrade

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

cicharka committed:
* separate get and set cluster version tasks
* introduce initial_cluster_version used to evaluate need of
  an upgrade

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

atsikham commented on docs/changelogs/CHANGELOG-1.3.md:
```diff
@@ -15,10 +15,14 @@
 - [#2644](https://github.com/epiphany-platform/epiphany/issues/2644) - Add validation to check hostnames for on-prem deployment
 - [#2703](https://github.com/epiphany-platform/epiphany/issues/2703) - Add tests for docker and kubelet cgroup driver
 - [#1076](https://github.com/epiphany-platform/epiphany/issues/1076) - Add sorting entries in the inventory file
+<<<<<<< HEAD
 - [#2768](https://github.com/epiphany-platform/epiphany/issues/2768) - Add posibility to provide custom hostnames
 - [#2785](https://github.com/epiphany-platform/epiphany/issues/2785) - Add configuration option to Keycloak for PROXY_ADDRESS_FORWARDING env. variable
 - [#2814](https://github.com/epiphany-platform/epiphany/issues/2814) - Add description how to enable TLS in Kibana
 - [#1076](https://github.com/epiphany-platform/epiphany/issues/2595) - Document connection protocols and ciphers
+=======
+- [#2702](https://github.com/epiphany-platform/epiphany/issues/2702) - Use state flag file in K8s upgrades
+>>>>>>> 743a2549 (Changelog updated)
```
Requires change

cicharka replied:
Already fixed

to-bar commented on ansible/playbooks/roles/upgrade/defaults/main.yml:
```diff
@@ -19,3 +19,7 @@ opendistro_for_elasticsearch:
 
   upgrade_state_file_path: /etc/elasticsearch/epicli-upgrade-started.state
+
+kubernetes:
+  upgrade_master_state_file_path: /var/lib/epiphany/upgrade/state/kubernetes-master-{{ ver }}.uncompleted
+  upgrade_node_state_file_path: /var/lib/epiphany/upgrade/state/kubernetes-node-{{ ver }}.uncompleted
```
```suggestion
kubernetes:
  upgrade_state_file_path: /var/lib/epiphany/upgrade/state/kubernetes-{{ ver }}.uncompleted
```
kubernetes_master and kubernetes_node roles are exclusive so how about common file?

cicharka replied:
Condition for the block **Upgrade nodes** [here](https://github.com/epiphany-platform/epiphany/pull/2766/files#diff-b642cd0642ad945c5f7f283861c694a6c5fb4218ed56c80eabf9df27fecd615bR81-R83) doesn't check if we are operating on master or node inventory_hostname, therefore it always enters block, creates flag file, skips upgrade, and then removes flag file again -> so even if the flag was present because of the master failure, it was removed inside upgrade **node block**. 

I will add all of your suggested changes, along with the additional change in [this condition](https://github.com/epiphany-platform/epiphany/pull/2766/files#diff-b642cd0642ad945c5f7f283861c694a6c5fb4218ed56c80eabf9df27fecd615bR81-R83), where we will check based on inventory hostname if we have to enter **node upgrade** block.

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
-      import_tasks: kubernetes/get-cluster-version.yml # sets cluster_version
+      import_tasks: kubernetes/get-cluster-version.yml
+
+    - name: Set initial_cluster_version fact # sets initial_cluster_version
```
```suggestion
    - name: Set initial_cluster_version fact
```
The fact name is just few lines below

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
+- name: Check upgrade state file for node
+  stat:
+    path: "{{ kubernetes.upgrade_node_state_file_path }}"
+  register: upgrade_node_state_file_status
```
```suggestion
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
+- name: Check upgrade state file for master
+  stat:
+    path: "{{ kubernetes.upgrade_master_state_file_path }}"
+  register: upgrade_master_state_file_status
```
```suggestion
- name: Check if upgrade state file exists
  stat:
    path: "{{ kubernetes.upgrade_state_file_path }}"
    get_attributes: false
    get_checksum: false
    get_mime: false
  register: k8s_upgrade_state_file_status
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
     - name: Upgrade masters
-      when: cluster_version is version('v' + version, '<=')
+      when: >
+        upgrade_master_state_file_status.stat.exists or
```
```suggestion
        k8s_upgrade_state_file_status.stat.exists or
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
       block:
+        - name: Create k8s upgrade state file for masters
+          copy:
+            dest: "{{ kubernetes.upgrade_master_state_file_path }}"
```
```suggestion
        - name: Create K8s upgrade state file on master node
          copy:
            dest: "{{ kubernetes.upgrade_state_file_path }}"
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +72,41 @@
+        - name: Remove k8s upgrade state file for masters
+          file:
+            path: "{{ kubernetes.upgrade_master_state_file_path }}"
+            state: absent
```
```suggestion
        - name: Remove K8s upgrade state file on master node
          file:
            path: "{{ kubernetes.upgrade_state_file_path }}"
            state: absent
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +72,41 @@
     - name: Upgrade nodes
-      when: kubelet_version is version('v' + version, '<=')
+      when: >
+        upgrade_node_state_file_status.stat.exists or
```
```suggestion
        k8s_upgrade_state_file_status.stat.exists or
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +72,41 @@
       block:
+        - name: Create k8s upgrade state file for nodes
+          copy:
+            dest: "{{ kubernetes.upgrade_node_state_file_path }}"
```
```suggestion
        - name: Create K8s upgrade state file on node
          copy:
            dest: "{{ kubernetes.upgrade_state_file_path }}"
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +72,41 @@
             - groups.kubernetes_node is defined
             - inventory_hostname in groups.kubernetes_node
 
+        - name: Remove k8s upgrade state file for nodes
+          file:
+            path: "{{ kubernetes.upgrade_node_state_file_path }}"
+            state: absent
+
```
```suggestion
        - name: Remove K8s upgrade state file on node
          file:
            path: "{{ kubernetes.upgrade_state_file_path }}"
            state: absent
```

atsikham commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
+  import_tasks: kubernetes/get-kubelet-version.yml
   delegate_to: "{{ groups.kubernetes_master[0] }}"
 
+- name: Set kubelet version as fact
```
As I see some tasks start with 'k8s |', and others aren't in the same place. Should be changed?

cicharka replied:
Honestly I am not aware that's the convention here, need to clarify.

cicharka replied:
I will change some of the following schema:
- if task includes or import another task for specific component I will add k8s
- if task is performing action on specific component/service I will add k8s
- if task is responsible for keeping the upgrade procedure ongoing, traces it and is a helper, I will leave it without k8s prefix
Do you agree with the proposed form?

atsikham commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
+        initial_cluster_version: >-
+          {{ (kubectl_cluster_version.stdout | from_yaml).serverVersion.gitVersion }}
+
+    - name: k8s/kubelet-cm | Include set-cluster-version.yml
```
`k8s/kubelet-cm` - is that name correct?

cicharka replied:
That name comes from [ansible/playbooks/roles/upgrade/tasks/kubernetes/patch-kubelet-cm.yml](https://github.com/epiphany-platform/epiphany/pull/2766/files/8039af8429e265e518dd817ad02b362ccca93d0f#diff-ee268bede56a4d59831f4e37a661445606a5f6669f046921cc627e5b0bbd1b51) so it is my mistake of not changing it here. I will fix it

atsikham commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -30,16 +72,41 @@
+            path: "{{ kubernetes.upgrade_node_state_file_path }}"
+            state: absent
+
+
```
Unnecessary newline. Could you apply formatting to be sure there is no such places and additional whitespaces?

cicharka replied:
Yes, I will fix it.

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,67 @@
+        initial_cluster_version: >-
+          {{ (kubectl_cluster_version.stdout | from_yaml).serverVersion.gitVersion }}
+
+    - name: k8s/kubelet-cm | Include set-cluster-version.yml
```
Not sure if we need this task here. If `cluster_version` is used somewhere before `patch-kubelet-cm.yml`, then I would just add the fact to the previous task.

cicharka replied:
We are using all of the facts (`version`, `major_version` and `minor_version`) that we are setting in `set-cluster-version.yml` -> Therefore it would be a code duplication if we will set it manually in previous task.

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,64 @@
   block:
     - name: Upgrade masters
-      when: cluster_version is version('v' + version, '<=')
+      when: >
```
Since now we have common flag file I would extend condition to run this block only on master nodes:
```
      when:
        - inventory_hostname in groups.kubernetes_master
        - k8s_upgrade_state_file_status.stat.exists
          or initial_cluster_version is version('v' + version, '<')
```

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes/gather-static-facts.yml:
```diff
@@ -8,6 +8,8 @@
         '.items[].metadata | select (.labels."k8s-app" == "canal"
                                   or .labels."k8s-app" == "calico-node"
                                   or .labels.app == "flannel") | .name'
+      args:
```
This fix is already in develop.

erzetpe commented:
/azp run

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,65 @@
+#   * identifying correct server version
+
+- name: k8s/master | Switch apiserver address to local
+  include_tasks: kubernetes/utils/set-local-apiserver.yml # sets kubectl_context_cluster
```
This file uses `kubectl` so to be on the safe side I would consider placing `k8s | Include wait-for-kube-apiserver.yml` task at the beginning, as we had before (it was the first task).

cicharka replied:
I moved `Wait for kube-apiserver` task at the beginning of the upgrade.

to-bar commented on ansible/playbooks/roles/upgrade/tasks/kubernetes.yml:
```diff
@@ -1,25 +1,65 @@
     - name: k8s | Include wait-for-kube-apiserver.yml
       import_tasks: kubernetes/utils/wait-for-kube-apiserver.yml
 
     - name: k8s | Include get-cluster-version.yml
```
I think we can simplify this flow. To decide about skipping upgrade we need only two common things (doesn't matter if it is master or worker node):
1) flag file existence
2) kubelet version

This means that probably we do not have to get cluster version here but only later, when condition is met and upgrade is not skipped.

Tasks `Create K8s upgrade state file on master node` and `Create K8s upgrade state file on node` have the same content so we can have one common task even not using YAML anchors. We could set fact node_role { master | worker }, then have:
```
- name: k8s/{{ node_role  }} | Create K8s upgrade state file
   copy:
      dest: "{{ kubernetes.upgrade_state_file_path }}"
      content: Upgrade started
      mode: u=rw,g=r,o=
```
The same with removal tasks.

cicharka replied:
As discussed with @to-bar and @seriva, we will not refactor that part for current release as it is more serious code interference with the upgrade of K8s.
Ticket for refactoring: #2844

cicharka committed:
move Wait for kube-apiserver task at the beginning of the upgrade

Signed-off-by: cicharka <arkadiusz.cichon@outlook.com>

erzetpe commented:
Tests passed on Ubuntu 18.04 LTS and Red Hat 7.8.
